sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,687 @@
+                which will cause a memory copy
+                if the given data is not C-contiguous.
+
+        y : Ignored
+            Not used, present here for API consistency by convention.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        self
+            Fitted estimator.
+        """
+        X = self._validate_data(X, accept_sparse='csr',
+                                dtype=[np.float64, np.float32],
+                                order='C', copy=self.copy_x,
+                                accept_large_sparse=False)
+
+        self._check_params(X)
+        random_state = check_random_state(self.random_state)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+        self._n_threads = _openmp_effective_n_threads()
+
+        # Validate init array
+        init = self.init
+
+        if hasattr(init, '__array__'):
+            init = check_array(init, dtype=X.dtype, copy=True, order='C')
+            self._validate_center_shape(X, init)
+
+        if self._algorithm == "full":
+            self._kmeans_single = _kmeans_single_lloyd
+            self._check_mkl_vcomp(X, X.shape[0])
+        else:
+            self._kmeans_single = _kmeans_single_elkan
+
+        if self.bisect_strategy == "child_biggest_sse":
+            _bisect_kmeans = self._bisect_by_biggest_child_sse
+
+        elif self.bisect_strategy == "largest_cluster":
+            _bisect_kmeans = self._bisect_largest_cluster
+        else:
+            # "biggest_sse"
+            _bisect_kmeans = self._bisect_by_biggest_sse
+
+        if self.verbose:
+            print("Running Bisecting K-Means with parameters:")
+            print(f"-> number of clusters: {self.n_clusters}")
+            print(f"-> number of centroid initializations: {self.n_init}")
+            print("-> relative tolerance: {:.4e}".format(self.tol))
+            print(f"-> bisect strategy: {self.bisect_strategy} \n")
+
+        x_squared_norms = row_norms(X, squared=True)
+
+        # Only assign to created centroid when n_clusters == 1
+        if self.n_clusters == 1:
+            self.cluster_centers_ = self._init_centroids(X, x_squared_norms,
+                                                         init, random_state,
+                                                         n_centroids=1)
+            warnings.warn("Bisection won't be performed - "
+                          "needs at least two clusters to run")
+        else:
+            self.cluster_centers_ = _bisect_kmeans(X, init, random_state,
+                                                   sample_weight)
+
+        # Since all clusters are calculated - label each data to valid center
+        self.labels_, self.inertia_ = _labels_inertia_threadpool_limit(
```

Comment:
It seems there are different possibilities for labeling (for both here and `predict`):
1. use the centroids as in KMeans, one sample being associated with the closest centroid over all centroids.
2. use the hierarchical bisecting tree, one sample being associated with the closest centroid recursively at each bisecting split.

You currently use (1), while pyspark uses (2). I think (2) makes more sense given the hierarchical structure of the algorithm. If you think (1) is still valuable, we could add a parameter to switch between the two methods. But then I would make (2) the default.

Reply:
Mostly because it may require overwriting `predict` and `score` quite unnecessary and I'm not sure if that wouldn't produce same result. Quite hard to me to say if it's thats valuable by now, so I'll change that to (2) option

Reply:
(1) might be slightly faster, but without benchmarks I can't tell it. I leave it with only (2) option, as it supposed to be.